File: backend/routes/stock.py
# backend/routes/stock.py
"""
Stock information endpoints (search and details).
Protected endpoints requiring JWT token verification.
"""
from fastapi import APIRouter, Depends
from logging import getLogger

from models.stock import StockSearchRequest, StockSearchResponse
from services.stock_service import StockService
from middleware.auth import verify_token


logger = getLogger(__name__)

router = APIRouter()
stock_service = StockService()


@router.post("/search")
async def search_stock(
    request: StockSearchRequest,
    current_user: dict = Depends(verify_token)
) -> StockSearchResponse:
    """
    Search stock by article code, location, or storage number.

    Filters results by activity code (act_code).
    Returns 0-N results matching any criterion.

    Request:
    {
        "art_code": "ABC123",     (optional)
        "stk_lieu": "A-01",       (optional)
        "stk_nosu": "SUPP-001",   (optional)
        "act_code": "BKS"         (required)
    }
    """
    logger.info(
        f"STOCK_SEARCH_REQUEST act_code={request.act_code} "
        f"art_code={request.art_code} stk_lieu={request.stk_lieu} stk_nosu={request.stk_nosu} "
        f"user={current_user.get('login')}"
    )

    response = stock_service.search_stock(request)

    logger.info(
        f"STOCK_SEARCH_RESPONSE act_code={request.act_code} "
        f"status={response.status} items_found={len(response.items)} "
        f"user={current_user.get('login')}"
    )

    return response


@router.get("/details/{sku}")
async def get_stock_details(
    sku: str,
    current_user: dict = Depends(verify_token)
) -> dict:
    """
    Get detailed stock information for a specific SKU.

    Path Parameters:
        sku: Product SKU

    Returns:
        dict: Detailed stock information
    """
    # TODO: Implement stock details logic
    return {
        "status": "success",
        "message": "Details endpoint (not yet implemented)",
        "sku": sku,
        "user": current_user.get("login"),
    }

File: backend-go/routes/stock.go
package routes

import (
	"fmt"
	"strings"

	"github.com/gofiber/fiber/v2"

	"stock-api/middleware"
	"stock-api/models"
	"stock-api/repositories"
	"stock-api/services"
)

// RegisterStock wires the stock endpoints onto the application.
func RegisterStock(app *fiber.App, repo *repositories.StockRepository) {
	svc := services.NewStockService(repo)
	app.Post("/api/stock/search", middleware.RequireAuth(), stockSearchHandler(svc))
	app.Get("/api/stock/details/:sku", middleware.RequireAuth(), stockDetailsHandler)
}

// stockSearchHandler godoc
// @Summary     Recherche stock
// @Description Recherche des articles en stock filtrés par activité. Au moins un critère parmi art_code, stk_lieu, stk_nosu est obligatoire.
// @Tags        stock
// @Accept      json
// @Produce     json
// @Security    BearerAuth
// @Param       body body models.StockSearchRequest true "Critères de recherche"
// @Success     200 {object} models.StockSearchResponse "Résultats (items peut être vide)"
// @Failure     401 {object} map[string]string "Token manquant ou invalide"
// @Failure     422 {object} map[string]string "Corps invalide ou act_code manquant"
// @Router      /api/stock/search [post]
func stockSearchHandler(svc *services.StockService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var req models.StockSearchRequest
		if err := c.BodyParser(&req); err != nil {
			return c.Status(fiber.StatusUnprocessableEntity).JSON(fiber.Map{
				"status":  "error",
				"message": "Invalid request body",
				"detail":  err.Error(),
			})
		}

		trimPtr(&req.ArtCode)
		trimPtr(&req.StkLieu)
		trimPtr(&req.StkNosu)
		req.ActCode = strings.TrimSpace(req.ActCode)

		if req.ActCode == "" {
			return c.Status(fiber.StatusUnprocessableEntity).JSON(fiber.Map{
				"status":  "error",
				"message": "act_code is required",
				"detail":  "act_code must not be empty",
			})
		}

		login, _ := c.Locals("login").(string)
		fmt.Printf("STOCK_SEARCH_REQUEST act_code=%s art_code=%s stk_lieu=%s stk_nosu=%s user=%s\n",
			req.ActCode, ptrStr(req.ArtCode), ptrStr(req.StkLieu), ptrStr(req.StkNosu), login)

		response, err := svc.SearchStock(req)
		if err != nil {
			if !hasAnyCriterion(req) {
				return c.Status(fiber.StatusUnprocessableEntity).JSON(fiber.Map{
					"status":  "error",
					"message": response.Message,
					"detail":  err.Error(),
				})
			}
			response.Message = "Search failed due to invalid criteria or database error"
			fmt.Printf("STOCK_SEARCH_RESPONSE act_code=%s status=%s items_found=%d user=%s\n",
				req.ActCode, response.Status, len(response.Items), login)
			return c.Status(fiber.StatusOK).JSON(response)
		}

		if len(response.Items) == 0 {
			response.Message = "No stock items found matching criteria"
		} else {
			response.Message = fmt.Sprintf("Found %d item(s)", len(response.Items))
		}

		fmt.Printf("STOCK_SEARCH_RESPONSE act_code=%s status=%s items_found=%d user=%s\n",
			req.ActCode, response.Status, len(response.Items), login)

		return c.Status(fiber.StatusOK).JSON(response)
	}
}

// stockDetailsHandler godoc
// @Summary     Détail stock par SKU
// @Description Retourne les informations détaillées d'un article (non encore implémenté)
// @Tags        stock
// @Produce     json
// @Security    BearerAuth
// @Param       sku path string true "Code article (SKU)"
// @Success     200 {object} map[string]string
// @Failure     401 {object} map[string]string "Token manquant ou invalide"
// @Router      /api/stock/details/{sku} [get]
func stockDetailsHandler(c *fiber.Ctx) error {
	sku := c.Params("sku")
	login, _ := c.Locals("login").(string)
	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"status":  "success",
		"message": "Details endpoint (not yet implemented)",
		"sku":     sku,
		"user":    login,
	})
}

func trimPtr(p **string) {
	if *p == nil {
		return
	}
	trimmed := strings.TrimSpace(**p)
	*p = &trimmed
}

func hasAnyCriterion(req models.StockSearchRequest) bool {
	return ptrStr(req.ArtCode) != "" || ptrStr(req.StkLieu) != "" || ptrStr(req.StkNosu) != ""
}

func ptrStr(p *string) string {
	if p == nil {
		return ""
	}
	return *p
}
